### ISSUE TITLE
`dead_code` false positive on `mod`+`lib`+`main`

### ISSUE THREAD

**Issue by Rudxain** [rust-lang/rust#124315]
> [!note]
> This is just [min-repro](https://en.wikipedia.org/wiki/Minimal_reproducible_example).
> It also happens when there's multiple identifiers, so long as those identifiers are unused by the module itself.

`src/utils.rs`:
```rust
pub fn ident<T>(v: T) -> T {
    v
}
```
`src/lib.rs`:
```rust
mod utils;
```
`src/main.rs`:
```rust
mod utils;
use utils::ident;

fn main() {
    println!("{}", ident(0));
}
```
`Cargo.toml`:
```toml
[package]
name = "undead"
version = "0.1.0"
edition = "2021"
```

I expected to see this happen: No warnings.

Instead, this happened:
```sh
$ cargo check
    Checking undead v0.1.0 (/home/rudxain/undead)
warning: function `ident` is never used
 --> src/utils.rs:1:8
  |
1 | pub fn ident<T>(v: T) -> T {
  |        ^^^^^
  |
  = note: `#[warn(dead_code)]` on by default

warning: `undead` (lib) generated 1 warning
    Finished dev [unoptimized + debuginfo] target(s) in 0.09s
```

### Meta

`rustc --version --verbose`:
```
rustc 1.77.2 (25ef9e3d8 2024-04-09)
binary: rustc
commit-hash: 25ef9e3d85d934b27d9dada2f9dd52b1dc63bb04
commit-date: 2024-04-09
host: x86_64-unknown-linux-gnu
release: 1.77.2
LLVM version: 17.0.6
```

Backtrace has same output as `cargo check`

**Comment by Rudxain** [comment]
I suspect this happens because `lib` takes shadows `main`, for some reason

**Comment by saethlin** [comment]
> hybrid crate

This is not a hybrid crate. This project contains two crates; this is documented at least in the Cargo book:

> If a package contains src/main.rs and src/lib.rs, it has two crates: a binary and a library, both with the same name as the package. A package can have multiple binary crates by placing files in the src/bin directory: each file will be a separate binary crate.

https://doc.rust-lang.org/book/ch07-01-packages-and-crates.html

**Comment by workingjubilee** [comment]
It is simply correct, the backtrace fn is never used when building the library.

This is a [cute trick that is not a good choice for actually shipping software](https://blog.axo.dev/2024/03/its-a-lib-and-a-bin).

Correct me if I am wrong, but I don't believe there is anything to fix here?

**Comment by fmease** [comment]
Indeed, this works as intended. I don't know how we could improve the diagnostics in such cases.
Technically speaking, it does already explicitly say that `ident` is only unused in the lib crate implying it's used in the bin crate:

> ```
> warning: `undead` (lib) generated 1 warning
> ```

If it were unused in both crates, the output would looks something like:

> ```
> warning: `undead` (lib) generated 1 warning
> warning: `undead` (bin "undead") generated 1 warning (1 duplicate)
> ```

If you have any concrete suggestions, we can reopen this issue as a diagnostic one.
Otherwise, this works as intended. Closing as such.
